ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+It does bring the benefit of avoiding namespace to subsequent calls to the client. However we feel
+like it is an unjustified and artificial limitation that will impact some multi-tenant applications.
+Besides this can always be brought in the shape of a `NamespacedClient` if there's a need for it.
+
+
+## Worker
+
+The worker interface allows SDK users to start processing registered workflow and activities. These
+will then be polled for on the specified namespaces & task queues and executed.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+
+worker_1 = Temporal::Worker.new(config, 'my-namespace', 'my-task-queue')
+worker_1.register_workflow(MyWorkflow) # uses 'MyWorkflow' as a name
+worker_1.register_activity(MyActivity) # uses 'MyActivity' as a name
+worker_1.start
+
+worker_2 = Temporal::Worker.new(config, 'my-namespace', 'another-task-queue')
+worker_2.register_workflow(MyWorkflow, name: 'my-workflow') # uses 'my-workflow' as a name
+worker_2.register_activity(MyActivity, name: 'my-activity') # uses 'my-activity' as a name
+worker_2.start
+
+# later
+worker_1.stop
+worker_2.stop
+```
+
+Notes:
+
+- This is not complete example of all configuration options for a worker
+- A worker will use the workflow/activity class name as a name unless it is explicitly provided (
+  `worker_2` demonstrated this approach)
+- Each worker will run in its own thread
+- A blocking `Worker#run` method will be available for convenience of running a single worker in
+  a process until SIGTERM or SIGINT is received
+- In case connection sharing is a concern we can create and memoize a connection on the instance of
+  Temporal::Configuration
+
+And here is the `Worker` interface:
+
+```ruby
+class Worker
+  # Create a new instance of a Worker
+  def self.new(config: Temporal::Configuration, namespace: String, task_queue: String, **options) -> Worker
+
+  # Register a workflow with an optional :name
+  def register_workflow(Class, name: nil) -> void
+
+  # Register an activity with an optional :name
+  def register_activity(Class, name: nil) -> void
+
+  # Start the worker and block until the process exits
+  def run() -> void
+
+  # Start the worker asynchronously
+  def start() -> void
+
+  # Gracefully stop the worker (wait for all running workflow/activities to finish)
+  def stop() -> void
+end
+```
+
+Internally each worker will create a thread for task polling as well as a thread pool of a given
+size (passed to the initializer as an optional `:threads` argument). Then each activity task will be
+dispatched to an available thread (via an in-memory queue). To make sure a polling worker always has
+processing capacity a polling request will only be dispatched if there is an available thread in the
+pool (in other words the pool will have no buffer).
+
+A very similar model will be used for workflow tasks, with an added complexity of wrapping each
+workflow in its own Fiber (inside a thread). This serves two purposes:
```

Comment:
What's the reason to have each workflow in it's own thread? A fiber-per-workflow makes a lot of sense but the thread should be unnecessary unless it's some kind of isolation mechanism. 

Ideally we can avoid it, Java uses a thread-per-everything and it has some downsides.

Reply:
@Sushisource The limitation is pretty straight forward — you can't run two fibers on one thread at a time. And while it's not a problem for running the workflow code itself, but that means you can't respond back to server and process a workflow at the same time unless it's done in multiple threads.

Fiber in this sense is more of an isolation mechanism that allows manual pre-empting, while Threads allow us to do more than one thing at a time

Reply:
@cretz @bergundy what about local activities? What if someone adds a network call into a local activity?

Reply:
Local activities run like regular activities from the SDK POV. They have the same concurrency concerns and are, from the SDK lang POV, unrelated to the concurrency concerns of the workflow.